libs/core/libs/shared/src/core_ops.rs

```diff
@@ -75,50 +79,50 @@ where
         Ok(File { id, parent, name, file_type, last_modified, last_modified_by, shares })
     }
 
-    // this variant used when we want to skip certain files e.g. when listing paths
-    pub fn resolve_and_finalize<I>(
-        &mut self, account: &Account, ids: I, public_key_cache: &mut LookupTable<Owner, String>,
-    ) -> SharedResult<Vec<File>>
-    where
-        I: Iterator<Item = Uuid>,
-    {
-        let mut user_visible_ids = Vec::new();
-        for id in ids {
-            if self.calculate_deleted(&id)? {
-                continue;
-            }
-            if self.in_pending_share(&id)? {
-                continue;
-            }
-            if self.link(&id)?.is_some() {
-                continue;
+    //finalizes one single id.
+    pub fn finalize(
+        &mut self, account: &Account, id: Uuid, public_key_cache: &mut LookupTable<Owner, String>,
+    ) -> SharedResult<File> {
+        let file;
+
+        let finalized = self.decrypt(&id, account, public_key_cache)?;
+
+        match finalized.file_type {
+            FileType::Document | FileType::Folder => file = finalized,
+            FileType::Link { target } => {
+                let mut target_file = self.decrypt(&target, account, public_key_cache)?;
+
+                target_file.id = finalized.id;
```

Comment:
you've done this already in `decrypt`. 

imo a refinement on the link discussion we had:

you're never going to change the id of a file. You're just going to change the parent sometimes, and you're going to change the parent if any link points to this file.

Reply:
right - if there's a link to this file, substitute its name and parent for the link's name and parent. I don't think the type of this file matters

Reply:
same goes for `finalize_all` below